Semantics.md

```diff
@@ -683,3 +683,22 @@ outside the range which rounds to an integer in range) traps.
 [future types]: FutureFeatures.md#more-table-operators-and-types
 [future large pages]: FutureFeatures.md#large-page-support
 [future ieee 754]: FutureFeatures.md#full-ieee-754-2008-conformance
+
+
+## Validation
+
+A module binary has to be _validated_ before it is compiled.
+Validation ensures that the module is well-defined and that its code cannot exhibit any undefined behavior.
+In particular, along with some runtime checks, it ensures that no program can access or corrupt memory it does not own.
+
+Validation of code is mostly defined in terms of type-checking the use of the operand stack.
+It sequentially checks for each that the expected operands can be popped from the stack, and tracks which new operands are pushed onto it. At the start of a function the stack is empty; at its end it has to match the return type of the function. In addition, instructions inside a `block` (or `loop` or `if`) cannot consume operands pushed outside; at the end of the block the remaining inner operands must match the block signature.
+
+A special case are instructions performing an unconditional control transfer (`br`, `br_table`, `return`, `unreachable`), because execution never proceeds after them.
+The stack after such an instruction is unconstrained, und thus said to be _polymorphic_.
+Instructions after an unconditional control transfer still need to type-check,
+but conceptually, values of any type can be popped off a polymorphic stack for the sake of checking consecutive instructions.
+A polymophic stack also matches any possible signature at the end of a block or function.
```

Comment:
From what I understand, the unconditional branch creates the polymorphic type, and thus is valid for any instructions.
If the instruction uses other values, not polymorphic, they must type-check.
Also, any following instructions still needs to correctly type check.
```
(module (func (i32.add (i64.const 1) (unreachable)))) ;; Invalid because [i64 poly] does not type-check for i32.add
(module (func (i32.add (i32.const 1) (unreachable)))) '' valid because [i32 poly] is valid for i32.add
(module (func (param i32)
  (unreachable) ;; code becomes unreachable after this
  (i32.add (i32.const 1) (get_local 0)) ;; still needs to do type-checking on this
))
```